common/app/common/dfp/PageskinAdAgent.scala

```diff
@@ -1,42 +1,51 @@
+import com.gu.commercial.display.MultipleValues
+import model.MetaData
 
 trait PageskinAdAgent {
 
   protected val environmentIsProd: Boolean
 
   protected def pageSkinSponsorships: Seq[PageSkinSponsorship]
 
-  private def findSponsorships(adUnitWithoutRoot: String, edition: Edition): Seq[PageSkinSponsorship] = {
+  // There are two forms of pageskins:
+  // - pageskins that target through ad unit (for pressed fronts)
+  // - pageskins that target through a keyword (for index page fronts)
+  private def findSponsorships(adUnitPath: String, metaData: MetaData, edition: Edition): Seq[PageSkinSponsorship] = {
 
-    if (PageSkin.isValidAdUnit(adUnitWithoutRoot)) {
-      val adUnitWithRoot = s"$dfpAdUnitGuRoot/$adUnitWithoutRoot"
+    val candidates = pageSkinSponsorships filter { sponsorship =>
+      sponsorship.editions.contains(edition) && !sponsorship.isR2Only }
 
-      def targetsAdUnitAndMatchesTheEdition(sponsorship: PageSkinSponsorship) = {
-        val adUnits = sponsorship.adUnits map (_.stripSuffix("/ng"))
-        adUnits.contains(adUnitWithRoot) &&
-          sponsorship.editions.contains(edition) &&
-          !sponsorship.isR2Only
-      }
-
-      pageSkinSponsorships filter { sponsorship =>
-        targetsAdUnitAndMatchesTheEdition(sponsorship)
-      }
+    if (metaData.isPressedPage) {
+      if (PageSkin.isValidAdUnit(adUnitPath)) {
+        candidates filter { sponsorship => sponsorship.adUnits.contains(adUnitPath) }
+      } else Seq.empty
     } else {
-      Seq.empty
+      val targetingMap = toMap(metaData.commercial.map(_.adTargeting(edition)).getOrElse(Set.empty))
+      val keywordTargeting = targetingMap.get("k") match {
+        case Some(values: MultipleValues) => values.values.toSeq
+        case _ => Seq.empty
+      }
+      candidates filter { sponsorship => sponsorship.keywords.intersect(keywordTargeting).nonEmpty }
     }
   }
 
   // The ad unit is considered to have a page skin if it has a corresponding sponsorship.
   // If the sponsorship is an adTest, it is only considered outside of production.
-  def hasPageSkin(adUnitWithoutRoot: String, edition: Edition): Boolean = {
-    findSponsorships(adUnitWithoutRoot, edition) exists (sponsorship => !(environmentIsProd && sponsorship.targetsAdTest))
+  def hasPageSkin(fullAdUnitPath: String, metaData: MetaData, edition: Edition): Boolean = {
+    if (metaData.isFront) {
+      findSponsorships(fullAdUnitPath, metaData, edition) exists (sponsorship =>
+        !(environmentIsProd && sponsorship.targetsAdTest))
+    } else false
   }
 
   // True if there is any candidate sponsorship for this ad unit. Used to decide when to render the out-of-page ad slot.
-  def hasPageSkinOrAdTestPageSkin(adUnitWithoutRoot: String, edition: Edition): Boolean = {
-    findSponsorships(adUnitWithoutRoot, edition).nonEmpty
+  def hasPageSkinOrAdTestPageSkin(fullAdUnitPath: String, metaData: MetaData, edition: Edition): Boolean = {
```

Comment:
Can this function be composed of two functions, hasPageSkin(...) and hasAdTestPageSkin(...) or is hasPageSkin in this context different to the one above?

Reply:
I've kept the original approach, which is increasing amounts of filtering based on shared `findSponsorships` function, because I found it quite readable.